resolver-dns/src/test/java/io/netty/resolver/dns/DnsNameResolverTest.java

```diff
@@ -1342,6 +1342,233 @@ int dnsRedirectPort(InetAddress server) {
+        final String domain = "netty.io";
+        final String hostname = "test.netty.io";
+        final String ns1Name = "ns1." + domain;
+        final InetSocketAddress ns1Address = new InetSocketAddress(
+                InetAddress.getByAddress(ns1Name, new byte[] { 10, 0, 0, 1 }),
+                DefaultDnsServerAddressStreamProvider.DNS_PORT);
+        final InetSocketAddress ns2Address = new InetSocketAddress(
+                InetAddress.getByAddress(ns1Name, new byte[] { 10, 0, 0, 2 }),
+                DefaultDnsServerAddressStreamProvider.DNS_PORT);
+        TestDnsServer redirectServer = new TestDnsServer(new HashSet<String>(Arrays.asList(hostname, ns1Name))) {
+            @Override
+            protected DnsMessage filterMessage(DnsMessage message) {
+                for (QuestionRecord record: message.getQuestionRecords()) {
+                    if (record.getDomainName().equals(hostname)) {
+                        message.getAdditionalRecords().clear();
+                        message.getAnswerRecords().clear();
+                        message.getAuthorityRecords().add(TestDnsServer.newNsRecord(domain, ns1Name));
+                        message.getAdditionalRecords().add(newARecord(ns1Address));
+                        message.getAdditionalRecords().add(newARecord(ns2Address));
+                        return message;
+                    }
+                }
+                return message;
+            }
+
+            private ResourceRecord newARecord(InetSocketAddress address) {
+                return TestDnsServer.newARecord(address.getHostName(), address.getAddress().getHostAddress());
+            }
+        };
+        redirectServer.start();
+        EventLoopGroup group = new NioEventLoopGroup(1);
+
+        final AtomicReference<List<InetSocketAddress>> redirectedRef = new AtomicReference<List<InetSocketAddress>>();
+        final DnsNameResolver resolver = new DnsNameResolver(
+                group.next(), new ReflectiveChannelFactory<DatagramChannel>(NioDatagramChannel.class),
+                NoopDnsCache.INSTANCE, NoopAuthoritativeDnsServerCache.INSTANCE,
+                NoopDnsQueryLifecycleObserverFactory.INSTANCE, 2000, ResolvedAddressTypes.IPV4_ONLY,
+                true, 10, true, 4096,
+                false, HostsFileEntriesResolver.DEFAULT,
+                new SingletonDnsServerAddressStreamProvider(redirectServer.localAddress()),
+                DnsNameResolver.DEFAULT_SEARCH_DOMAINS, 0, true) {
+
+            @Override
+            protected List<InetSocketAddress> uncachedRedirectDnsServerList(
+                    String hostname, List<InetSocketAddress> nameServers) {
+                redirectedRef.set(nameServers);
+                return super.uncachedRedirectDnsServerList(hostname, nameServers);
+            }
+        };
+
+        try {
+            assertNotNull(resolver.resolveAll(hostname).await().cause());
```

Comment:
Could we at least check exception type and message?

Reply:
Sure...